File: 413 - Arithmetic Slices (Medium)/C++/solution.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& A) {
        if (A.size() < 3) return 0; // There must be three terms for an arithmetic sequence to exist.

        int current = 0;    // The current length of the arithmetic series.
        int sum = 0;        // The total number of arithmetic series.

        for (int i = 2; i < A.size(); i++) { // Iterate through every number beyond i = 1;
            if (A[i - 2] - A[i - 1] == A[i - 1] - A[i]) { // The index is part of an arith. series.
                current += 1;   // Add  to the length of the current arithmetic series.
                sum += current; // Add the current length to the sum.
            }
            else { // Otherwise, we found the end of an arithmetic series.
                current = 0; // Reset the current length.
            }
        }

        return sum; // Sum is our answer.
    }
};

File: 413 - Arithmetic Slices (Medium)/Python/solution.py
class Solution(object):
    def numberOfArithmeticSlices(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        # There must be three terms for an arithmetic sequence to exist.
        if len(A) < 3: return 0

        dp, offset = [0] * len(A), 1 # Initialize our dp array and offset.

        for i in range(2, len(A)): # Iterate through everything past i = 1
            if A[i - 1] - A[i - 2] == A[i] - A[i - 1]: # The number is part of an arith. series.
                dp[i] = dp[i - 1] + offset  # Copy the value forward and add the offset to it.
                offset += 1                 # Increment the offset.
            else: # Otherwise, reset the offset and copy the value forward.
                dp[i] = dp[i - 1]
                offset = 1


        return dp[-1] # The last index of the dp array is our answer.
